onnxruntime/core/providers/acl/math/gemm.h

```diff
@@ -0,0 +1,268 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) 2019, NXP Semiconductor, Inc. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/framework/op_kernel.h"
+#include "core/util/math.h"
+#include "core/util/math_cpuonly.h"
+#include "core/providers/cpu/math/gemm.h"
+#include "core/providers/acl/acl_execution_provider.h"
+
+// ACL
+#include "arm_compute/core/TensorInfo.h"
+#include "arm_compute/runtime/TensorAllocator.h"
+#include "arm_compute/runtime/Allocator.h"
+#include "arm_compute/runtime/PoolManager.h"
+#include "arm_compute/runtime/BlobLifetimeManager.h"
+#include "arm_compute/runtime/MemoryManagerOnDemand.h"
+
+//NEON
+#include "arm_compute/runtime/NEON/functions/NEGEMM.h"
+#include "arm_compute/runtime/NEON/functions/NETranspose.h"
+
+#define GEMM_CPU
+#define CACHE_TRANSPOSED_DATA
+#undef DEBUG_GEMM
+
+namespace onnxruntime {
+namespace acl {
+
+typedef struct {
+    std::shared_ptr<arm_compute::NEGEMM> layer;
+    std::shared_ptr<arm_compute::Tensor> a, b, c, d;
+    std::shared_ptr<arm_compute::MemoryManagerOnDemand> mm_layer;
+} ACLNEGEMM;
+
+typedef std::map<OpKernel*, ACLNEGEMM>::iterator GEMMLayersIterator;
+
+#ifndef GEMM_CPU
+static std::shared_ptr<arm_compute::MemoryManagerOnDemand> createMemoryManager()
+{
+    auto lifetime_mgr  = std::make_shared<arm_compute::BlobLifetimeManager>();
+    auto pool_mgr      = std::make_shared<arm_compute::PoolManager>();
+    auto mm            = std::make_shared<arm_compute::MemoryManagerOnDemand>(lifetime_mgr, pool_mgr);
+
+    return mm;
+}
+#endif
+
+template <typename T>
+class Gemm : public onnxruntime::Gemm<T> {
+public:
+  Gemm(const OpKernelInfo& info) : onnxruntime::Gemm<T>(info) {
+    int64_t temp;
+
+    ORT_ENFORCE(info.GetAttr<int64_t>("transA", &temp).IsOK());
+    trans_A_ = temp == 0 ? CblasNoTrans : CblasTrans;
+    ORT_ENFORCE(info.GetAttr<int64_t>("transB", &temp).IsOK());
+    trans_B_ = temp == 0 ? CblasNoTrans : CblasTrans;
+
+    ORT_ENFORCE(info.GetAttr<float>("alpha", &alpha_).IsOK());
+    ORT_ENFORCE(info.GetAttr<float>("beta", &beta_).IsOK());
+
+#ifdef DEBUG_GEMM
+#ifndef GEMM_CPU
+    printf("Gemm ACL\n");
+#else
+    printf("Gemm CPU\n");
+#endif
+#endif
+  }
+
+#ifndef GEMM_CPU
```

Comment:
What's the purpose of the #ifndef GEMM_CPU #else blocks? Was it for debug/testing?
If we build with ACL, wouldn't we always want the ACL implementation?

Reply:
GEMM implementation from ACL EP if functional but for the moment it is below CPU EP performance. I propose to keep it disabled until we manage to improve its performance.